docs/api-reference/shadertools/shader-modules/gpu-picking.md

```diff
@@ -0,0 +1,188 @@
+
+## Under the Hood
+
+Depending on the structure of the shader, each object can either correspond to an `instance` or a group of vertexes.
+
+An additional consideration 
+
+li
+
+The `gpuPicking` modules supports:
+- picking of object indexes
+- highlighting of objects
+- pick a specific *instance* in an instanced draw call
+- highlight all fragments of an *instance* based on its picking color
+- pick "group of primitives" with the same picking color in non-instanced draw-calls
+- highlight "group of primitives" with the same picking color in non-instanced draw-calls
+
+Highlighting allows applications to specify a picking color corresponding to an object that need to be highlighted and the highlight color to be used.
+
+## About GPU based picking
+
+GPU based picking has a couple of significant advantage over CPU-based picking:
+- GPU-based picking is a picking technique that can be performed entirely on the GPU, meaning that it is very performant, especially when picking is done every frame.
+
+- can be added to any existing shaders 
+- and is independent of the structure of the input geometry or rendering without requiring any additional picking logic to that shader, beyond calling one function in the vertex shader and one function in the fragment shader.
+
+Note that GPU-based picking does comes with some limitations:
+- Picking occluding objects require re-rendering and discarding the already picked objects.
+- On WebGL-specific: the read back of the picking data from the picking texture can only be done synchronously, causing a GPU pipeline stall, which can defeat some of the performance advantages.
+
+Traditional 3d frameworks often support CPU-based picking. While luma.gl does not include an CPU-based picking algorithms, CPU based picking techniques do have advantages: 
+ - They can often provide precise intersection points on objects and they are better at handling picking of multiple objects, especially for objects that are occluded. 
+ - However CPU based picking techniques are slower and can require more data on the CPU or they may need to be customized to the structure of the input data.
+
+
+## Usage
+
+In your vertex shader, your inform the picking module what object we are currently rendering by supplying a picking color, perhaps from an attribute.
+
+```ts
+attribute vec3 aPickingColor;
+main() {
+  picking_setPickingColor(aPickingColor);
+  ...
+}
+```
+
+In your fragment shader, you simply apply (call) the `picking_filterColor` filter function at the very end of the shader. This will return the normal color, or the highlight color, or the picking color, as appropriate.
+
+```ts
+main() {
+  gl_FragColor = ...
+  gl_FragColor = picking_filterPickingColor(gl_FragColor);
+}
+```
+
+In your fragment shader, you simply apply (call) the `picking_filterPickingColor` filter function at the very end of the shader. This will return the normal color, or the highlight color, or the picking color, as appropriate.
```

Comment:
Duplicate paragraph

Reply:
Yep, the docs need work, found them quite difficult to write.